datalake/src/main/java/com/sequenceiq/datalake/service/recovery/SdxRecoverySelectorService.java

```diff
@@ -0,0 +1,66 @@
+package com.sequenceiq.datalake.service.recovery;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+
+import com.sequenceiq.cloudbreak.api.endpoint.v4.stacks.response.recovery.RecoveryStatus;
+import com.sequenceiq.cloudbreak.common.exception.BadRequestException;
+import com.sequenceiq.datalake.entity.SdxCluster;
+import com.sequenceiq.sdx.api.model.SdxRecoverableResponse;
+import com.sequenceiq.sdx.api.model.SdxRecoveryRequest;
+import com.sequenceiq.sdx.api.model.SdxRecoveryResponse;
+
+@Service
+/*
+ * Recovers an SDX cluster from a failure during Resize or Upgrade.
+ *
+ * Chooses the appropriate RecoveryService to use based on the state of the SDX cluster.
+ */
+public class SdxRecoverySelectorService {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(SdxRecoverySelectorService.class);
+
+    private  final List<RecoveryService> services;
+
+    @Autowired
+    public SdxRecoverySelectorService(List<RecoveryService> services) {
+        this.services = services;
+    }
+
+    public SdxRecoveryResponse triggerRecovery(SdxCluster sdxCluster, SdxRecoveryRequest sdxRecoveryRequest) {
+
+        List<String> responseReasons = new ArrayList<>();
+
+        for (RecoveryService recoveryService : services) {
+            SdxRecoverableResponse recoverableResponse = recoveryService.validateRecovery(sdxCluster);
+            responseReasons.add(recoverableResponse.getReason());
+            if (recoverableResponse.getStatus().recoverable()) {
+                return recoveryService.triggerRecovery(sdxCluster, sdxRecoveryRequest);
+            }
+        }
+
+        String reasons = String.join(",", responseReasons);
+        LOGGER.debug("Cluster is not in a recoverable state with message(s): {}", reasons);
+        throw new BadRequestException(reasons);
+
+    }
+
+    public SdxRecoverableResponse validateRecovery(SdxCluster sdxCluster) {
+
+        SdxRecoverableResponse sdxRecoverableResponse = null;
```

Comment:
if there are multiple non-recoverable responses, then this logic selects either of them, it might be better to return all the response reasons collected together in the response.

Reply:
The api still expects only a single reason in the response, so without modifying that, concatenating the reasons is the only way to make this possible.

Reply:
Yes, I get that, my issue here is that if a client calls `validateRecovery` explicitly, then only one of the errors is returned, not the aggregate, concatenated error.

Reply:
It will at least now return all of the messages together. I think a change to the API would be overkill at this point